test/MusicStore.Test/ShoppingCartControllerTest.cs

```diff
@@ -0,0 +1,173 @@
+        {
+            // Arrange
+            var sessionFeature = new SessionFeature()
+            {
+                Session = CreateTestSession(),
+            };
+
+            var httpContext = new DefaultHttpContext();
+            httpContext.SetFeature<ISessionFeature>(sessionFeature);
+            httpContext.Session.SetString("Session", "CartId_A");
+
+            var controller = new ShoppingCartController()
+            {
+                DbContext = _serviceProvider.GetRequiredService<MusicStoreContext>(),
+            };
+            controller.ActionContext.HttpContext = httpContext;
+
+            // Act
+            var result = await controller.Index();
+
+            // Assert
+            var viewResult = Assert.IsType<ViewResult>(result);
+            Assert.NotNull(viewResult.ViewData);
+            Assert.Null(viewResult.ViewName);
+
+            var model = Assert.IsType<ShoppingCartViewModel>(viewResult.ViewData.Model);
+            Assert.Equal(0, model.CartItems.Count);
+            Assert.Equal(0, model.CartTotal);
+        }
+
+        [Fact]
+        public async Task Index_ReturnsCartItems_WhenItemsInCart()
+        {
+            // Arrange
+            var cartId = "CartId_A";
+            var sessionFeature = new SessionFeature()
+            {
+                Session = CreateTestSession(),
+            };
+
+            var httpContext = new DefaultHttpContext();
+            httpContext.SetFeature<ISessionFeature>(sessionFeature);
+            httpContext.Session.SetString("Session", cartId);
+
+            var dbContext = _serviceProvider.GetRequiredService<MusicStoreContext>();
+            var cartItems = CreateTestCartItems(
+                cartId,
+                itemPrice: 10,
+                numberOfItem: 5);
+            dbContext.AddRange(cartItems.Select(n => n.Album).Distinct());
+            dbContext.AddRange(cartItems);
+            dbContext.SaveChanges();
+
+            var controller = new ShoppingCartController()
+            {
+                DbContext = dbContext,
+            };
+            controller.ActionContext.HttpContext = httpContext;
+
+            // Act
+            var result = await controller.Index();
+
+            // Assert
+            var viewResult = Assert.IsType<ViewResult>(result);
+            Assert.NotNull(viewResult.ViewData);
+            Assert.Null(viewResult.ViewName);
+
+            var model = Assert.IsType<ShoppingCartViewModel>(viewResult.ViewData.Model);
+            Assert.Equal(5, model.CartItems.Count);
+            Assert.Equal(5 * 10, model.CartTotal);
+        }
+
+        private static ISession CreateTestSession()
+        {
+
+            return new DistributedSession(
+                new LocalCache(new MemoryCache(new MemoryCacheOptions())),
+                "sessionId_A",
+                idleTimeout: TimeSpan.MaxValue,
+                tryEstablishSession: () => true,
+                loggerFactory: new NullLoggerFactory(),
```

Comment:
Could you instantiate a Logger for all the tests in general so that individual components can continue to emit logs. https://github.com/aspnet/MusicStore/blob/dev/test/E2ETests/SmokeTests.cs#L23